### ISSUE TITLE
Use defvar-keymap via Compat

### ISSUE THREAD

**Issue by minad** [oantolin/embark#586]
I've been thinking about using it for Embark and maybe removing the `embark-define-keymap` macro in favor of `defvar-keymap`.

_Originally posted by @oantolin in https://github.com/oantolin/orderless/issues/131#issuecomment-1407762588_

**Comment by minad** [comment]
(Had to open this issue such that it will not get lost, see https://github.com/oantolin/orderless/issues/131#issuecomment-1407764139.)

**Comment by minad** [comment]
Thanks! We could in principle also port back elp or eww functions. But we have to draw a line somewhere, so I would rather avoid this. Also Compat won't port back commands.

As a bad example Compat also provides a backport of a dired function, which I don't think is a good fit. Unfortunately it got added at the request of Jonas before I took over. But I have to discuss with him if we should drop this again.

https://github.com/emacs-compat/compat/blob/41460f01f0a808c9e8855f41417ded8cf804f442/compat-27.el#L569

(See also the compat-call mechanism as documented in the Compat manual.)

**Comment by minad** [comment]
But then Dired could be considered core functionality.

Here is another borderline case:

https://github.com/emacs-compat/compat/commit/01a8a9dc90dbd4d4a69b906e01488a63d003d65d

But these ERT functions are useful to test Compat itself, so they are maybe sufficiently low level/library level. On the other hand it is testing code, so maybe not useful in general? 

Generally functions which live in a non-preloaded feature or which need compat-call should not be backported lightheartedly. What do you think?

**Comment by minad** [comment]
The docs and wiki should also be updated with respect to embark-define-keymap. I commented on the commit, besides that it looks good to me!

**Comment by oantolin** [comment]
> The docs and wiki should also be updated with respect to embark-define-keymap.

Oh, right. I'll update them tomorrow.

**Comment by minad** [comment]
@oantolin Do you have an opinion about the limitations of Compat, as mentioned in my comments
https://github.com/oantolin/embark/issues/586#issuecomment-1407990171 and https://github.com/oantolin/embark/issues/586#issuecomment-1407999602?

**Comment by oantolin** [comment]
It is hard to know where to stop with compat. Maybe it means something that without looking too much into the library I instinctively felt it would contain `defvar-keymap`, which it did, but that it would not have either `eww-download-callback` or `elp-restore-package`, which it didn't. I guess you can take that as evidence that there is some agreement on what constitutes core Emacs functionality and that people might expect compat to limit itself to that.

But specifically that dired function and ert stuff? That seems hard to decide for me. On the one hand dired is an application that runs on Emacs, it's not infrastructure in the same way list manipulation or key bindings are. But, as you said dired is arguably core Emacs functionality. I often think it would be nice if Emacs moved a bunch of packages to GNU ELPA, but even when I'm imagining that, I still expect dired to ship with Emacs. I mean, there are files and there are directories and I expect to edit both without installing additional packages from GNU ELPA.

The ert stuff I think would be OK for a different reason: it is infrastructure or at least scaffolding around infrastructure, it is purely language related and having it encourages good practice.

I don't know, I can really see the issue going either way on those two cases. Maybe I'm convincing myself that ert is more appropriate than dired, but including both or excluding both also seen reasonable.

**Comment by minad** [comment]
> I often think it would be nice if Emacs moved a bunch of packages to GNU ELPA, but even when I'm imagining that, I still expect dired to ship with Emacs. I mean, there are files and there are directories and I expect to edit both without installing additional packages from GNU ELPA.

Yes, Dired should and would always ship with Emacs. Like you I would prefer if stuff moved out of core, in contrast to what is happening right now, where more and more stuff is added.

However I still believe that Dired functionality should not be part of Compat, since it is an application on top of Emacs.

> The ert stuff I think would be OK for a different reason: it is infrastructure or at least scaffolding around infrastructure, it is purely language related and having it encourages good practice.

I feel the same way.

> Maybe I'm convincing myself that ert is more appropriate than dired, but including both or excluding both also seen reasonable.

Yes, the same with me. My main gripe is currently the `dired-get-marked-files` function, which I feel was added prematurely to Compat. Note that the Dired function is also only available via `compat-call` and I think that `:extended` and `:feature` functions should have to pass a higher threshold. The function specifies both `:extended` and `:feature`, which is like the worst possible combination.

https://github.com/emacs-compat/compat/blob/41460f01f0a808c9e8855f41417ded8cf804f442/compat-27.el#L572-L573
